src/item_factory.cpp

```diff
@@ -1495,6 +1506,53 @@ void Item_factory::load_item_category(JsonObject &jo)
     }
 }
 
+void Item_factory::load_migration( JsonObject &jo )
+{
+    migration m;
+    m.id = jo.get_string( "id" );
+    assign( jo, "replace", m.replace );
+    assign( jo, "destroy", m.destroy );
+    assign( jo, "flags", m.flags );
+    assign( jo, "charges", m.charges );
+    assign( jo, "contents", m.contents );
+
+    migrations[ jo.get_string( "id" ) ] = m;
+}
+
+itype_id Item_factory::migrate_id( const itype_id& id )
+{
+    auto iter = migrations.find( id );
+    if( iter != migrations.end() ) {
+        if( iter->second.destroy ) {
+            return "null";
+        }
+        if( !iter->second.replace.empty() ) {
+            return iter->second.replace;
+        }
+    }
+    return id;
+}
+
+void Item_factory::migrate_item( const itype_id& id, item& obj )
+{
+    auto iter = migrations.find( id );
+    if( iter != migrations.end() ) {
+        if( iter->second.destroy ) {
+            obj.type = item::nullitem();
```

Comment:
Outright destroying the item makes the whole loading code more complicated (you had to introduce several `is_null` checks) and player will report this as a bug ("item just vanished upon loading my save!").

Why not just convert it to a standard note item ("obsolete item"), maybe with a proper description explaining that this item had been removed from the game. This avoids the null check (still a valid item), and allows the player to discover that change.